src/plugins/irc/irc-server.c

```diff
@@ -4313,75 +4283,188 @@ irc_server_check_certificate_fingerprint (struct t_irc_server *server,
     char **fingerprints;
     int i, rc, algo;
-    size_t size_bits, size_bytes;
-
-    for (i = 0; i < IRC_FINGERPRINT_NUM_ALGOS; i++)
-    {
-        fingerprint_server[i] = NULL;
-    }
+    size_t j, size_bytes, pkbuf_len;
+    gnutls_pubkey_t pk;
 
     /* split good_fingerprints */
     fingerprints = weechat_string_split (good_fingerprints, ",", NULL,
                                          WEECHAT_STRING_SPLIT_STRIP_LEFT
                                          | WEECHAT_STRING_SPLIT_STRIP_RIGHT
                                          | WEECHAT_STRING_SPLIT_COLLAPSE_SEPS,
                                          0, NULL);
+
     if (!fingerprints)
         return 0;
 
     rc = 0;
+    pkbuf = NULL;
+    pkbuf_len = 0;
 
-    for (i = 0; fingerprints[i]; i++)
+    for (algo = 0; algo < IRC_FINGERPRINT_NUM_ALGOS; algo++)
     {
-        size_bits = strlen (fingerprints[i]) * 4;
-        size_bytes = size_bits / 8;
-
-        algo = irc_server_fingerprint_search_algo_with_size (size_bits);
-        if (algo < 0)
-            continue;
+        certfpstr[algo] = NULL;
+        keyfpstr[algo] = NULL;
+    }
 
-        if (!fingerprint_server[algo])
+    /* attempt to extract and fingerprint the certificate's public key */
+    if (gnutls_pubkey_init (&pk) != GNUTLS_E_SUCCESS)
+    {
+        weechat_printf (
+            server->buffer,
+            _("%sgnutls: failed to initialize public key object"),
+            weechat_prefix ("error"));
+    }
+    else
+    {
+        if (gnutls_pubkey_import_x509 (pk, certificate, 0)
+            != GNUTLS_E_SUCCESS)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to import certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (gnutls_pubkey_export (pk, GNUTLS_X509_FMT_DER, NULL,
+                     &pkbuf_len) != GNUTLS_E_SHORT_MEMORY_BUFFER)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (!pkbuf_len)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (!(pkbuf = malloc (pkbuf_len)))
+        {
+            weechat_printf (
+                server->buffer,
+                _("%s%s: not enough memory (%s)"),
+                weechat_prefix ("error"), IRC_PLUGIN_NAME,
+                "fingerprint");
+        }
+        else if (gnutls_pubkey_export (pk, GNUTLS_X509_FMT_DER, pkbuf,
+                     &pkbuf_len) != GNUTLS_E_SUCCESS)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
```

Comment:
This whole public key extraction might deserve to be a separate function, if one doesn't already exist in GnuTLS.

Reply:
It does exist in GnuTLS, in that `gnutls_pubkey_export` is all you get. This scaffolding around it is necessary to prepare for that export.

There is no `gnutls_x509_crt_get_*` function that will hand you the `SubjectPublicKeyInfo` structure; the closest is `gnutls_x509_crt_get_pk_rsa_raw` which hands you the modulus and exponent separately (and would hardcode us to RSA).

I can move it to a separate function on request, but they would still be heavily tied together...